### PR TITLE
Perform bitreverse using AVX512 GFNI for i32 and i64.

### PR DESCRIPTION
**Pull request by shamithoke**
Currently, the lowering operation for bitreverse using Intel AVX512 GFNI supports only byte vectors 

Extend the operation to i32 and i64.

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -31040,17 +31044,63 @@ static SDValue LowerBITREVERSE_XOP(SDValue Op, SelectionDAG &DAG) {
   return DAG.getBitcast(VT, Res);
 }
 
+static auto createBSWAPShuffleMask(EVT VT) {
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31047]: Style: We avoid to use `auto` in function definition. And it's better to pass a `SmallVectorImpl<int> &ShuffleMask` to avoid copy.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -1845,6 +1845,10 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
     setOperationAction(ISD::UMULO, MVT::v64i8, Custom);
 
     setOperationAction(ISD::BITREVERSE, MVT::v64i8,  Custom);
+    if (Subtarget.hasGFNI())
+      for (auto VT : {MVT::i32, MVT::v4i32, MVT::v8i32, MVT::v16i32, MVT::i64,
+                      MVT::v2i64, MVT::v4i64, MVT::v8i64})
+        setOperationAction(ISD::BITREVERSE, VT, Custom);
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:1851]: Traditionally we need to change `hasVLX` for 128/256-bits vectors. I think we can loose the check now @RKSimon ?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 3]
```diff
@@ -31040,17 +31044,63 @@ static SDValue LowerBITREVERSE_XOP(SDValue Op, SelectionDAG &DAG) {
...
+  SDLoc DL(Op);
+
+  auto HasGFNI = Subtarget.hasGFNI();
+  auto ScalarType = VT.getScalarType();
+
+  if (HasGFNI && ((ScalarType == MVT::i32) || (ScalarType == MVT::i64))) {
+    if (VT.isVector()) {
+      SmallVector<int, 16> BSWAPMask = createBSWAPShuffleMask(VT);
+      EVT ByteVT =
+          EVT::getVectorVT(*DAG.getContext(), MVT::i8, BSWAPMask.size());
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31070]: This can be `MVT`

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 4]
```diff
@@ -31040,17 +31044,63 @@ static SDValue LowerBITREVERSE_XOP(SDValue Op, SelectionDAG &DAG) {
...
-  SDLoc DL(Op);
+  assert(VT.getScalarType() == MVT::i8 &&
+         "Only byte vector BITREVERSE supported");
+
+  // Split v64i8 without BWI so that we can still use the PSHUFB lowering.
+
+  if (Subtarget.hasXOP() && !VT.is512BitVector())
+    return LowerBITREVERSE_XOP(Op, DAG);
+
+  assert(Subtarget.hasSSSE3() && "SSSE3 required for BITREVERSE");
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31103]: Doesn't it duplicate line 31090?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 5]
```diff
@@ -31040,17 +31044,63 @@ static SDValue LowerBITREVERSE_XOP(SDValue Op, SelectionDAG &DAG) {
...
   MVT VT = Op.getSimpleValueType();
+  SDValue In = Op.getOperand(0);
+  SDLoc DL(Op);
+
+  auto HasGFNI = Subtarget.hasGFNI();
+  auto ScalarType = VT.getScalarType();
+
+  if (HasGFNI && ((ScalarType == MVT::i32) || (ScalarType == MVT::i64))) {
+    if (VT.isVector()) {
+      SmallVector<int, 16> BSWAPMask = createBSWAPShuffleMask(VT);
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31068]: Why not just create a BSWAP and let the legalizer deal with this instead of manually creating the shuffle?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 6]
```diff
@@ -31040,17 +31044,63 @@ static SDValue LowerBITREVERSE_XOP(SDValue Op, SelectionDAG &DAG) {
...
+
+  return ShuffleMask;
+}
+
 static SDValue LowerBITREVERSE(SDValue Op, const X86Subtarget &Subtarget,
                                SelectionDAG &DAG) {
   MVT VT = Op.getSimpleValueType();
+  SDValue In = Op.getOperand(0);
+  SDLoc DL(Op);
+
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31062]: For scalar cases, we'd be better off pulling the scalar code out of LowerBITREVERSE_XOP and just putting it at the top of LowerBITREVERSE - all x86 BITREVERSE custom lowering code uses vector instructions.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 2 above)
Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:1851]: I have the feeling we'd be better off improving SSSE3/AVX/AVX512 coverage to always handle all vector types (not just vXi8) - and that would mean we could drop all the GFNI setOperationAction cases apart from the scalar types (which should be pulled out of the AVX512 case so Tremont can use them properly).

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 1 above)
Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31047]: I have updated the function definition like you suggested. However, I think, because of RVO (Return Value Opitmization/Copy elision), the copy would not have been made.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 3 above)
Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: Done.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 5 above)
Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31068]: I am not sure if I entirely follow your suggestion, but I had tried extracting elements and applying BSWAP on each element (every 4 bytes for integer). This, however, resulted in 4 byte-swaps. 
If this is not what you meant, can you please elaborate a bit more?

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 4 above)
Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: Yes, it was a mistake. I have removed the duplicate lines.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 5 above)
Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31068]: Please see c95febcb4008c40a2b514bd6e2b56e0aa17457a3 - this should handle it for you

Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31068]: Thank you!

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 6 above)
Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31062]: Actually, that's not going to work as LowerBITREVERSE_XOP  folds the bswap into the VPPERM shuffle - we need our own implementation:
```cpp
  if (!VT.isVector()) {
    MVT VecVT = MVT::getVectorVT(VT, 128 / VT.getSizeInBits());
    SDValue Res = VT == MVT::i8 ? In : DAG.getNode(ISD::BSWAP, DL, VT, In);
    Res = DAG.getNode(ISD::SCALAR_TO_VECTOR, DL, VecVT, Res);
    Res = DAG.getBitcast(MVT::v16i8, Res);
    Res = DAG.getNode(ISD::BITREVERSE, DL, MVT::v16i8, Res);
    Res = DAG.getBitcast(VecVT, Res);
    return DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, VT, Res,
                       DAG.getIntPtrConstant(0, DL));
  }
```

**Review by RKSimon — CHANGES_REQUESTED** [review]
Pull the GFNI setOperationAction calls out of the hasAVX512 case so any GFNI target can make use of it

**Comment by RKSimon** [comment]
@shamithoke reverse-ping?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 7]
```diff
@@ -31332,6 +31337,21 @@ static SDValue LowerBITREVERSE(SDValue Op, const X86Subtarget &Subtarget,
   if (VT.is256BitVector() && !Subtarget.hasInt256())
     return splitVectorIntUnary(Op, DAG, DL);
 
+  // Lower i32/i64 to GFNI as i32/i64 -> Convert to vector (V = v16i32/v8i64) -> vXi8 BITREVERSE -> V[0] -> BSWAP
+  if (Subtarget.hasGFNI() && !VT.isVector()) {
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31341]: No need for GFNI check - this is generic vectorization lowering

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 8]
```diff
@@ -31332,6 +31337,21 @@ static SDValue LowerBITREVERSE(SDValue Op, const X86Subtarget &Subtarget,
...
+
+    assert ((VT.getScalarType() == MVT::i32) || (VT.getScalarType() == MVT::i64));
+
+    auto ScalarType = VT.getScalarType();
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31345]: (style) Don't use auto

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 9]
```diff
@@ -31332,6 +31337,21 @@ static SDValue LowerBITREVERSE(SDValue Op, const X86Subtarget &Subtarget,
...
+    auto CastTo = ScalarType == MVT::i32 ? MVT::v4i32 : MVT::v2i64;
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31346]: Just do this generically (so it can work for other scalar types if we ever want to)...
```
MVT SVT = VT.getScalarType();
MVT VecVT = MVT::getVectorVT(SVT, 128 / SVT.getSizeInBits());
```

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 10]
```diff
@@ -31332,6 +31337,21 @@ static SDValue LowerBITREVERSE(SDValue Op, const X86Subtarget &Subtarget,
   if (VT.is256BitVector() && !Subtarget.hasInt256())
     return splitVectorIntUnary(Op, DAG, DL);
 
+  // Lower i32/i64 to GFNI as i32/i64 -> Convert to vector (V = v16i32/v8i64) -> vXi8 BITREVERSE -> V[0] -> BSWAP
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31340]: Please can you rephrase this?

Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31340]: Simplified it. Does this work?

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 7 above)
Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31341]: I've changed it, but then it would mean that any call to lower_bitreverse for i32/i64 (for non-GFNI targets) will get lowered as PSHUFB lookups. I thought since this PR explicitly handles GFNI target, so I will put a stringent condition.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 8 above)
Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31345]: Done.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 9 above)
Comment (shamithoke) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:31346]: Done.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 11]
```diff
@@ -31332,6 +31337,21 @@ static SDValue LowerBITREVERSE(SDValue Op, const X86Subtarget &Subtarget,
   if (VT.is256BitVector() && !Subtarget.hasInt256())
     return splitVectorIntUnary(Op, DAG, DL);
 
+  // Lower i32/i64 to GFNI as vXi8 BITREVERSE + BSWAP
+  if (!VT.isVector()) {
+
+    assert ((VT.getScalarType() == MVT::i32) || (VT.getScalarType() == MVT::i64));
+
+    MVT SVT = VT.getScalarType();
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:31345]: Remove SVT - its guaranteed to be the same as VT

**Review by RKSimon — COMMENTED** [review]
Please can you run clang-format on your changes?

**Comment by shamithoke** [comment]
> Please can you run clang-format on your changes?

Done

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by shamithoke** [comment]
> LGTM - cheers

Thank you!

**Comment by shamithoke** [comment]
> LGTM - cheers

But I do not think I have commit access.
Can you please merge this one for me?

**Comment by RKSimon** [comment]
@shamithoke I forgot to ask - why didn't you extend this for i8/i16 cases?

**Comment by shamithoke** [comment]
> @shamithoke I forgot to ask - why didn't you extend this for i8/i16 cases?

I needed the bitreverse for i32/i64 on AVX, so I just handled those case. 
Let me make a PR to add i8 and i16.
